tools/purge_executions.py

```diff
@@ -58,101 +56,110 @@ def do_register_cli_opts(opts, ignore_errors=False):
     else:
         global DELETED_COUNT
         DELETED_COUNT += 1
 
     try:
         ActionExecution.delete(execution_db)
-    except Exception as e:
-        print('Exception deleting Execution model: %s, exception: %s',
-              execution_db, str(e))
+    except:
+        LOG.exception('Exception deleting Execution model: %s',
+                      execution_db)
     else:
-        try:
-            LiveAction.delete(liveaction_db)
-        except Exception as e:
-            print('Zombie LiveAction left in db: %s. Exception: %s',
-                  liveaction_db, str(e))
+        if liveaction_db:
+            try:
+                LiveAction.delete(liveaction_db)
+            except:
+                LOG.exception('Zombie LiveAction left in db: %s.', liveaction_db)
+
 
+def _should_delete(execution_db, action_ref, timestamp, purge_incomplete=False):
+    if not purge_incomplete and execution_db.status in IN_PROGRESS_STATES:
+        return False
 
-def _purge_executions(timestamp=None, action_ref=None):
+    if action_ref != '':
+        return (execution_db.liveaction['action'] == action_ref and
+                execution_db.start_timestamp < timestamp)
+    else:
+        return execution_db.start_timestamp < timestamp
+
+
+def purge_executions(timestamp=None, action_ref=None, purge_incomplete=False):
     if not timestamp:
-        print('Specify a valid timestamp to purge.')
+        LOG.error('Specify a valid timestamp to purge.')
         return
 
     if not action_ref:
         action_ref = ''
 
-    print('Purging executions older than timestamp: %s' %
-          timestamp.strftime('%Y-%m-%dT%H:%M:%S.%fZ'))
-
-    def should_delete(execution_db):
-        if action_ref != '':
-            return (execution_db.liveaction['action'] == action_ref and
-                    execution_db.start_timestamp < timestamp)
-        else:
-            return execution_db.start_timestamp < timestamp
+    LOG.info('Purging executions older than timestamp: %s' %
+             timestamp.strftime('%Y-%m-%dT%H:%M:%S.%fZ'))
 
     # XXX: Think about paginating this call.
-    filters = {'start_timestamp__lt': isotime.parse(timestamp)}
+    filters = {'end_timestamp__lt': isotime.parse(timestamp)}
     executions = ActionExecution.query(**filters)
-    executions_to_delete = filter(should_delete, executions)
-    print('#### Total number of executions to delete: %d' % len(executions_to_delete))
+    executions_to_delete = [execution for execution in executions
+                            if _should_delete(
+                                execution_db=execution, action_ref=action_ref,
+                                timestamp=timestamp,
+                                purge_incomplete=purge_incomplete)]
+    LOG.info('#### Total number of executions to delete: %d' % len(executions_to_delete))
 
     # Purge execution and liveaction models now
     for execution_db in executions_to_delete:
```

Comment:
We have to assume this will be a large set. Do you think a bulk delete if you choose not to use delete_by_query?